api/xapi.rb

```diff
@@ -0,0 +1,171 @@
+require 'trackler'
+require 'exercism'
+require "bugsnag"
+require './config/build_id'
+require './config/bugsnag'
+require 'exercism/homework'
+require_relative 'xapi/helpers'
+
+module Xapi
+  EXERCISM_URL = "https://github.com/exercism/exercism.io"
+
+  class Restoration
+    attr_reader :implementation, :files
+    def initialize(implementation:, files:)
+      @implementation = implementation
+      @files = files
+    end
+
+    def self.for(submission)
+      track = Trackler.tracks[submission.track_id]
+      return unless track.exists?
+
+      implementation = track.implementations[submission.slug]
+      return unless implementation.exists?
+
+      files = implementation.files.merge submission.solution
+
+      Restoration.new(implementation: implementation, files: files)
+    end
+  end
+
+  class App < Sinatra::Base
+    configure do
+      after { ActiveRecord::Base.connection.close }
+    end
+
+    configure do
+      set :root, Exercism.relative_to_root('api', 'xapi')
+      enable :raise_errors
+      use Rack::Session::Cookie,
+        key: 'rack.session',
+        path: '/',
+        expire_after: 2_592_000,
+        secret: ENV.fetch('SESSION_SECRET') { 'Need to know only.' }
+    end
+
+    error Sinatra::NotFound do
+      halt 404, { error: "endpoint '#{request.path}' not found." }.to_json
+    end
+
+    error 500 do
+      puts $ERROR_INFO.inspect
+      Bugsnag.auto_notify($ERROR_INFO, nil, request)
+      msg = "So sorry! We've been notified of the error and will investigate."
+      { error: msg }.to_json
+    end
+
+    helpers ::Xapi::Helpers::Guards
+    helpers ::Xapi::Helpers::Finders
+
+    get '/tracks' do
+      pg :tracks, locals: {
+        tracks: Trackler.tracks,
+        todos: Trackler.todos,
+      }
+    end
+
+    get '/tracks/:id' do |id|
+      track = find_track(id)
+
+      pg :track, locals: {
+        track: track,
+        todos: Trackler.todos,
+      }
+    end
+
+    # Put 'restore' before /v2/exercise/:track_id
+    # otherwise we get 'unknown track restore'.
+    get '/v2/exercises/restore' do
+      require_key
+
+      exercises = current_user.exercises.order(:language, :slug)
+      solutions = exercises.map { |ex| ex.submissions.last }.compact
+      restorations = solutions.map { |solution| Restoration.for(solution) }.compact
+
+      pg :restore, locals: { things_to_restore: restorations }
+    end
+
+    get '/v2/exercises' do
+      require_key
+
+      track_ids = params[:tracks].to_s.split(",").map {|s| s.strip}
+      if track_ids.empty?
+        track_ids = Trackler.tracks.select(&:active?).map(&:id)
+      end
+      track_ids = track_ids & Trackler.tracks.map(&:id)
+
+      implementations = []
+
+      track_ids.each do |track_id|
+        track = Trackler.tracks[track_id]
+        slugs = slugs_by_track_id[track_id]
+        # pretend they already solved hello-world if they've
+        # solved anything at all.
+        slugs << 'hello-world' unless slugs.empty?
```

Comment:
possibly extract 

```ruby
slugs << 'hello-world' unless slugs.empty?
next_slug = (track.implementations.map(&:slug) - slugs).first
if !!next_slug
  implementation = track.implementations[next_slug]
  if implementation.exists?
    implementations << implementation
  end
end
```
into its own method so that you don't have to repeat the logic in two routes?
* get '/v2/exercises'
* get '/v2/exercises/:track_id'